staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -121,52 +118,26 @@ func runServer(server *http.Server, network string, stopCh <-chan struct{}) (int
 		return 0, fmt.Errorf("invalid listen address: %q", ln.Addr().String())
 	}
 
-	lock := sync.Mutex{} // to avoid we close an old listener during a listen retry
+	// Stop the server by closing the listener
 	go func() {
 		<-stopCh
-		lock.Lock()
-		defer lock.Unlock()
 		ln.Close()
 	}()
 
 	go func() {
-		defer utilruntime.HandleCrash()
-
-		for {
-			var listener net.Listener
-			listener = tcpKeepAliveListener{ln.(*net.TCPListener)}
-			if server.TLSConfig != nil {
-				listener = tls.NewListener(listener, server.TLSConfig)
-			}
-
-			err := server.Serve(listener)
-			glog.Errorf("Error serving %v (%v); will try again.", server.Addr, err)
-
-			// listen again
-			func() {
-				lock.Lock()
-				defer lock.Unlock()
-				for {
-					time.Sleep(15 * time.Second)
-
-					ln, err = net.Listen(network, server.Addr)
-					if err == nil {
-						return
-					}
-					select {
-					case <-stopCh:
-						return
-					default:
-					}
-					glog.Errorf("Error listening on %v (%v); will try again.", server.Addr, err)
-				}
-			}()
-
-			select {
-			case <-stopCh:
-				return
-			default:
-			}
+		var listener net.Listener
+		listener = tcpKeepAliveListener{ln.(*net.TCPListener)}
+		if server.TLSConfig != nil {
+			listener = tls.NewListener(listener, server.TLSConfig)
+		}
+
+		err := server.Serve(listener)
+
+		stopRequested := strings.Contains(err.Error(), "use of closed network connection")
```

Comment:
why not keep the old logic with the channel? This looks fishy to compare error texts.